### PR TITLE
Fast multiplication

### LINKED ISSUE #43: Speed higher-order multiplications
The vec trick still seems to work for higher-order Kronecker products, by means of recursion, e.g.

```julia
A = B = C = randn(10, 10);
x = rand(1000)

kronecker(A, B, C) * x
```
works.

I think however that there is still room for performance improvement, especially when all the components of the product are square and for higher-order products. In this case, the array containing the partial results have the same size, so a new one should not be allocated every step. Especially [TensorOperations.jl](https://github.com/Jutho/TensorOperations.jl) seems much faster. 

Either have a specialized method that allocates a single array to store the results or use an external package?

### PR DESCRIPTION
**Pull request by emerali**
Resolves #43 

TODO:

- [x] Get tests working
- [ ] ~~Implement fast multiplication for KroneckerPower~~ shelved, see below
- [x] Reduce code duplication
- [x] Add methods for fast `\`
- [x] Reduce (some) allocations
- [ ] ~~Add a specialized method for when all factors in a KroneckerProduct are the same size (some rough benchmarks i did seem to indicate there are performance gains to be achieved)~~ shelved for a later PR

### CODE REVIEW

**Comment by emerali** [comment]
I'm playing around with applying multiplication by a (square) KroneckerPower using a generalization of FWHT and it seems to be giving me very similar performance to the (square) shuffle algorithm, so I'm not sure if it's necessary right now to add a separate implementation

File: src/base.jl
Comment (emerali) [reply src/base.jl:92]: good catch!

File: src/kroneckersum.jl
Comment (elisno) [inline src/kroneckersum.jl:68]: I'm guessing you want to use this implementation _exclusively_ for fast multiplications with Kronecker sums 
(in `mul!(out::AbstractVecOrMat, K::AbstractKroneckerSum, x::AbstractVecOrMat`) . :heart_eyes:

Suppose you have the higher order products/sums:
```julia
K1 = A ⊗ B ⊗ C
K2 = A ⊕ B ⊕ C
```
Currently, you would get `getallmatrices(K1) == getallmatrices(K2) == (A, B, C)` and run the risk of forgetting 
 what the original subtype of  the `GeneralizedKroneckerProduct` you are working with.

Although `getallmatrices` isn't exported, this behavior may need to be documented if someone wants to use it themselves.

I imagine that this would be a problem with dispatching `mul!` for (sum)mixed products like `A ⊕ B ⊗ C ⊕ D`, but that case may be outside the scope of this PR.

Comment (emerali) [reply src/kroneckersum.jl:68]: yeah that is a problem...

i'll change it to `getallfactors` for products/powers and `getallsummands` for kronsums, unless you have any better names?

Comment (elisno) [reply src/kroneckersum.jl:68]: Maybe shorten each to `getfactors` and `getsummands`?

Comment (emerali) [reply src/kroneckersum.jl:68]: i think those could end up being confused with `getmatrices` in that users may expect it to return a 2-tuple

**Comment by emerali** [comment]
Ok, I think things are in a decent shape as is. I'd like someone to look at the type signatures, specifically considering eltypes of arrays as I'm not too well versed in dealing with eltype promotion.

Otherwise, I don't have anything else to add (to this PR at least). There's still a few optimizations that can be done (ex. simplifying the mul/ldiv functions for when all factors are the same size), which I'll experiment with locally before opening up another PR.

**Review by MichielStock — APPROVED** [review]

**Comment by MichielStock** [comment]
Amazing work! I played around with it, and it now really works well with higher-order products. 
I don't see how to improve. For me it is ok to merge into master, unless you want someone else (maybe @elisno?) to take a look?

**Comment by elisno** [comment]
This is very impressive!
I think most of the issues I brought up in the review don't need to be addressed in this PR, but I'd appreciate hearing your opinions.

It should be fine to merge this to master.

**Review by elisno — APPROVED** [review]

File: benchmark/benchmarks.jl
Comment (emerali) [reply benchmark/benchmarks.jl]: fixed

File: benchmark/benchmarks.jl
Comment (emerali) [reply benchmark/benchmarks.jl]: fixed

File: src/multiply.jl
Comment (emerali) [reply src/multiply.jl:39]: basically, yeah

should i prepend a `_` to these as well since they aren't "public" functions either?

File: src/vectrick.jl
Comment (emerali) [reply src/vectrick.jl:58]: Main reason i just outright deleted those was because they weren't exported at all, so i'm not sure if it could be considered a breaking change

If I understand correctly, these methods were written to "skip" identity matrices when applying the vectrick recursively. Not sure if they're needed anymore given that vectrick is now only used in the case of 2 kronecker factors. For this reason I didn't compare the runtimes.

Now that you mention it, maybe these methods could give a slight speed boost in the 2-factor case, but I'll have to leave that for later.

File: src/multiply.jl
Comment (elisno) [reply src/multiply.jl:39]: Most of these functions generated programmatically, I'd consider to be internal.
It would be more consistent to prepend with an underscore.

File: src/vectrick.jl
Comment (emerali) [reply src/vectrick.jl:147]: ok now i'm starting to wonder why the mixed-product property isn't being applied here

i'm pretty sure that was implemented somewhere given that there are tests for it...

File: src/vectrick.jl
Comment (emerali) [reply src/vectrick.jl:117]: added a short comment on this

File: src/kroneckersum.jl
Comment (emerali) [reply src/kroneckersum.jl:69]: i would say that is the correct/expected behavior, no? i dont think users would expect a kronecker product to have summands

Comment (elisno) [reply src/kroneckersum.jl:69]: > i dont think users would expect a kronecker product to have summands

I agree. In this case, I wonder if we should return an error?

Comment (emerali) [reply src/kroneckersum.jl:69]: hmm, I'm slightly averse to throwing an error in this case since the error is more of a "warning", i think

File: src/vectrick.jl
Comment (elisno) [reply src/vectrick.jl:148]: Currently, the mixed-product property only applies to `Base.:*`
https://github.com/MichielStock/Kronecker.jl/blob/master/src/base.jl#L331

Thanks for spotting this issue! 👍

Comment (emerali) [reply src/vectrick.jl:148]: It seems the current implementation errors out when dealing with kronecker products whose factors are incompatibly-sized.

Is this behavior something we wanna keep?

Comment (elisno) [reply src/vectrick.jl:148]: Are you talking about the implementation for `Base.:*`, that checks the relevant dimensions of the factors? I don't see an issue with that one.

What behavior did you have in mind? Can you share some code examples that shows this unexpected "behavior"?

Comment (emerali) [reply src/vectrick.jl:148]: basically just have it collect the second argument in the case of incompatible sizes

from a user's perspective I dont think I'd expect an error in this case, but it's likely not an issue that'd pop up for me personally (i'm usually dealing with kronecker products of a bunch of 2x2 matrices)

File: src/kroneckersum.jl
Comment (elisno) [reply src/kroneckersum.jl:69]: A warning would be much better, especially when dealing with mixtures of products and sums, e.g.

```julia
K = A ⊕ B ⊗ C ⊕ D
getallfactors(K)  # Tuple{KroneckerSum{Float64,KroneckerSum{Float64,Array{Float64,2},Kronecker.KroneckerProduct{Float64,Array{Float64,2},Array{Float64,2}}},Array{Float64,2}}}
getallsummands(K) # Tuple{Array{Float64,2},Kronecker.KroneckerProduct{Float64,Array{Float64,2},Array{Float64,2}},Array{Float64,2}}
```
I'm not saying that this needs to be solved right now, as I'm not sure which operator should take precedence in this case.

Comment (emerali) [reply src/kroneckersum.jl:69]: yeah, I think we should open an issue for that to discuss it further (once this PR is merged ofc)

File: src/vectrick.jl
Comment (emerali) [reply src/vectrick.jl:148]: as it is now, should I implement `mul!` functions for the `AbstractKroneckerProduct * AbstractKroneckerProduct` case in this PR? or just leave things as is? I'll still let it error out in the case of mismatched factor sizes ofc.

Comment (elisno) [reply src/vectrick.jl:132]: Now that I think of it, `mul!(C::AbstractMatrix, K1::AbstractKroneckerProduct, K2::AbstractKroneckerProduct)` would just be equivalent to `collect!(C, K1 * K2)` in https://github.com/MichielStock/Kronecker.jl/blob/master/src/base.jl#L296

I'd just put this into a new issue after merging this PR, leaving things as is. Perhaps we should then discuss the usage of each method as it is reflected in the documentation.

Comment (emerali) [reply src/vectrick.jl:132]: oops i already pushed those lol

should i revert that commit or just leave it?

Comment (elisno) [reply src/vectrick.jl:132]: I scanned quickly over the latest changes and they look reasonable.
However, it think it belongs in a separate PR due to the size of the current one (there we could add some more tests as well).

You can revert the commit, stash and apply to a fresh branch for a new PR after this one is merged.

Comment (emerali) [reply src/vectrick.jl:148]: all right, should be good now

**Comment by emerali** [comment]
is there anything else that needs to be done for this PR?

**Comment by elisno** [comment]
I think everything's covered for now.

**Comment by MichielStock** [comment]
awesome!

**Pull request merged**
